### PR TITLE
log query on fatal error

### PR DESCRIPTION
**Pull request by save-my-heart**
Changelog category (leave one):
- Improvement

Since server may crash before writing log to `system.query_log` and `system.query_thread_log`, so we can not know which query had caused this when log level is `WARNING` or `ERROR`. 
We can get query detail in signal handler, then the daemon will log this query if possible.

### CODE REVIEW

File: base/daemon/BaseDaemon.cpp [hunk 1]
```diff
@@ -81,13 +81,17 @@ static void call_default_signal_handler(int sig)
 
 static constexpr size_t max_query_id_size = 127;
 
+// equal to setting 'log_queries_cut_to_length'
+static constexpr size_t max_query_size = 100000;
```
Comment (azat) [inline base/daemon/BaseDaemon.cpp:85]: This is not correct, since you cannot write such many bytes to pipe atomically.

Comment (azat) [reply base/daemon/BaseDaemon.cpp:85]: See https://github.com/ClickHouse/ClickHouse/blob/2966e9aac4c39dbee6cd1aa3cde3607eaec1b45f/src/Common/TraceSender.cpp#L37-L40 for the reference

File: base/daemon/BaseDaemon.cpp [hunk 2]
```diff
@@ -81,13 +81,17 @@ static void call_default_signal_handler(int sig)
...
+
 static const size_t signal_pipe_buf_size =
     sizeof(int)
     + sizeof(siginfo_t)
     + sizeof(ucontext_t)
     + sizeof(StackTrace)
     + sizeof(UInt32)
     + max_query_id_size + 1    /// query_id + varint encoded length
+    + max_query_size + 1
```
Comment (azat) [inline base/daemon/BaseDaemon.cpp:94]: It will not be 1 byte for `max_query_size` since it is >255

File: base/daemon/BaseDaemon.cpp (hunk 1 above)
Comment (azat) [reply base/daemon/BaseDaemon.cpp:85]: End even though there is no SA_NODEFER, this handler is used for multiple signals, so if multiple of them will be arrived then the information in pipe can be overlapped, yes this is not that likely, but does it worth it?

Comment (save-my-heart) [reply base/daemon/BaseDaemon.cpp:85]: > End even though there is no SA_NODEFER, this handler is used for multiple signals, so if multiple of them will be arrived then the information in pipe can be overlapped, yes this is not that likely, but does it worth it?

Thanks, maybe it is just only for special use.

**Comment by alexey-milovidov** [comment]
By the way, signal handler is sleeping for 10 seconds and system logs are flushed every 7 seconds
\- exactly to let queries to be logged in case of crash.

**Comment by save-my-heart** [comment]
@alexey-milovidov In my cases, clickhouse server crashed when calling `Interpreter`, which queries had not been written to system logs. I could only get stack trace and query id through server error log, so I want to find out which query caused the problem. 
Although I can easily change the log level to catch that query, I would like if there is a better way.

**Comment by save-my-heart** [comment]
@azat Another question, seems current `signal_pipe_buf_size` is > `PIPE_BUF` on some platforms。
https://github.com/ClickHouse/ClickHouse/blob/7a4743662734e8b4b3709ade38e09ea220ef1733/base/daemon/BaseDaemon.cpp#L84-L91

**Pull request closed without merging**